[0.9.3-rc3] restart influxdb causing panic

influxdb 0.9.3rc3
centos 7.1
sometime influxdb stop write the point,and I restart the influxdb 
restart influxdb with high load causing panic
```
[run] 2015/08/26 12:57:48 Signal received, initializing clean shutdown...
[run] 2015/08/26 12:57:48 Waiting for clean shutdown...
[snapshot] 2015/08/26 12:57:48 snapshot listener closed
[tcp] 2015/08/26 12:57:48 cluster service accept error: network connection closed
[http] 2015/08/26 12:57:48 172.17.0.30 - root [26/Aug/2015:12:57:48 +0800] POST /write?db=metrics&p=%5BREDACTED%5D&precision=s&u=root HTTP/1.1 204 0 - Ruby ff6ba6dc-4bae-11e5-a2ea-000000000000 9.527186ms
[shard-precreation] 2015/08/26 12:57:48 Precreation service terminating
panic: assignment to entry in nil map

goroutine 662224 [running]:
github.com/influxdb/influxdb/tsdb.(*Store).CreateShard(0xc2080341a0, 0xc2441cfd8f, 0x7, 0xc20f810370, 0x7, 0x1, 0x0, 0x0)
        /home/philip/build2/src/github.com/influxdb/influxdb/tsdb/store.go:90 +0x727
github.com/influxdb/influxdb/cluster.func·002(0x1, 0x1, 0xc234d5c000, 0x6a, 0x80)
        /home/philip/build2/src/github.com/influxdb/influxdb/cluster/points_writer.go:245 +0x1c4
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
        /home/philip/build2/src/github.com/influxdb/influxdb/cluster/points_writer.go:271 +0x30b

goroutine 1 [select]:
main.(*Main).Run(0xc20806bb00, 0xc20800a010, 0x4, 0x4, 0x0, 0x0)
        /home/philip/build2/src/github.com/influxdb/influxdb/cmd/influxd/main.go:101 +0xe06
main.main()
        /home/philip/build2/src/github.com/influxdb/influxdb/cmd/influxd/main.go:42 +0xdc
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -36,6 +36,8 @@ With this release InfluxDB is moving to Go 1.5.
 - [#3980](https://github.com/influxdb/influxdb/pull/3980): 'service stop' waits until service actually stops. Fixes issue #3548.
 - [#4016](https://github.com/influxdb/influxdb/pull/4016): Shutdown Graphite UDP on SIGTERM.
 - [#4034](https://github.com/influxdb/influxdb/pull/4034): Rollback bolt tx on mapper open error
+- [#3848](https://github.com/influxdb/influxdb/issues/3848): restart influxdb causing panic
+- [#3881](https://github.com/influxdb/influxdb/issues/3881): panic: runtime error: invalid memory address or nil pointer dereference
 
 ## v0.9.3 [2015-08-26]
 

--- a/cluster/service.go
+++ b/cluster/service.go
@@ -108,7 +108,7 @@ func (s *Service) Close() error {
 
 	// Shut down all handlers.
 	close(s.closing)
-	// s.wg.Wait() // FIXME(benbjohnson)
+	s.wg.Wait()
 
 	return nil
 }

--- a/cmd/influxd/run/server.go
+++ b/cmd/influxd/run/server.go
@@ -387,7 +387,8 @@ func (s *Server) Close() error {
 		s.Listener.Close()
 	}
 
-	// Close services to any inflight requests can be stopped
+	// Close services to allow any inflight requests to complete
+	// and prevent new requests from being accepted.
 	for _, service := range s.Services {
 		service.Close()
 	}

--- a/cmd/influxd/run/server.go
+++ b/cmd/influxd/run/server.go
@@ -382,23 +382,32 @@ func (s *Server) Open() error {
 func (s *Server) Close() error {
 	stopProfile()
 
+	// Close the listener first to stop any new connections
 	if s.Listener != nil {
 		s.Listener.Close()
 	}
-	if s.MetaStore != nil {
-		s.MetaStore.Close()
+
+	// Close services to any inflight requests can be stopped
+	for _, service := range s.Services {
+		service.Close()
 	}
-	if s.TSDBStore != nil {
-		s.TSDBStore.Close()
-	}
-	if s.HintedHandoff != nil {
-		s.HintedHandoff.Close()
-	}
+
 	if s.Monitor != nil {
 		s.Monitor.Close()
 	}
-	for _, service := range s.Services {
-		service.Close()
+
+	if s.HintedHandoff != nil {
+		s.HintedHandoff.Close()
+	}
+
+	// Close the TSDBStore, no more reads or writes at this point
+	if s.TSDBStore != nil {
+		s.TSDBStore.Close()
+	}
+
+	// Finally close the meta-store since everything else depends on it
+	if s.MetaStore != nil {
+		s.MetaStore.Close()
 	}
 
 	close(s.closing)

--- a/tsdb/store.go
+++ b/tsdb/store.go
@@ -37,6 +37,7 @@ type Store struct {
 
 	EngineOptions EngineOptions
 	Logger        *log.Logger
+	closing       chan struct{}
 }
 
 // Path returns the store's root path.
@@ -66,6 +67,12 @@ func (s *Store) ShardN() int {
 func (s *Store) CreateShard(database, retentionPolicy string, shardID uint64) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
+
+	select {
+	case <-s.closing:
+		return fmt.Errorf("closing")
+	default:
+	}
 
 	// shard already exists
 	if _, ok := s.shards[shardID]; ok {
@@ -297,6 +304,8 @@ func (s *Store) Open() error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	s.closing = make(chan struct{})
+
 	s.shards = map[uint64]*Shard{}
 	s.databaseIndexes = map[string]*DatabaseIndex{}
 
@@ -352,6 +361,8 @@ func (s *Store) Close() error {
 			return err
 		}
 	}
+	close(s.closing)
+	s.closing = nil
 	s.shards = nil
 	s.databaseIndexes = nil
 

